subsys/bluetooth/host/gatt.c

```diff
@@ -100,12 +100,17 @@ static ssize_t read_appearance(struct bt_conn *conn,
 
 static struct bt_gatt_attr gap_attrs[] = {
 	BT_GATT_PRIMARY_SERVICE(BT_UUID_GAP),
-#if CONFIG_BT_DEVICE_NAME_MAX > 0
+#if defined(CONFIG_BT_DEVICE_NAME_DYNAMIC)
+#if defined(CONFIG_BT_DEVICE_NAME_GATT_WRITABLE)
 	/* Require pairing for writes to device name */
 	BT_GATT_CHARACTERISTIC(BT_UUID_GAP_DEVICE_NAME,
 			       BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE,
 			       BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT,
 			       read_name, write_name, bt_dev.name),
+#else
+	BT_GATT_CHARACTERISTIC(BT_UUID_GAP_DEVICE_NAME, BT_GATT_CHRC_READ,
+			       BT_GATT_PERM_READ, read_name, NULL, bt_dev.name),
```

Comment:
Perhaps changing read_name to rely on bt_get_name would be better here, that way we don't need the if statement for dynamic just for writable.